gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/kafka/KeyValueEventObjectReporter.java

```diff
@@ -0,0 +1,173 @@
+
+    Config config = builder.config.get();
+    Config pusherConfig = ConfigUtils.getConfigOrEmpty(config, PUSHER_CONFIG).withFallback(config);
+    String pusherClassName =
+        ConfigUtils.getString(config, PUSHER_CLASS, PusherUtils.DEFAULT_KEY_VALUE_PUSHER_CLASS_NAME);
+    this.pusher =
+        PusherUtils.getKeyValuePusher(pusherClassName, builder.brokers, builder.topic, Optional.of(pusherConfig));
+    this.closer.register(this.pusher);
+
+    randomKey = String.valueOf(new Random().nextInt(100));
+    if (config.hasPath(PUSHER_KEYS)) {
+      List<String> keys = Splitter.on(",").omitEmptyStrings().trimResults().splitToList(config.getString(PUSHER_KEYS));
+      this.keys = Optional.of(keys);
+    } else {
+      log.warn("Key not assigned from config. Please set it with property {}",
+          ConfigurationKeys.METRICS_REPORTING_EVENTS_KAFKAPUSHERKEYS);
+      log.warn("Using generated number " + randomKey + " as key");
+    }
+  }
+
+  @Override
+  public void reportEventQueue(Queue<GobblinTrackingEvent> queue) {
+    log.info("Emitting report using KeyValueEventObjectReporter");
+
+    List<Pair<String, GenericRecord>> events = Lists.newArrayList();
+    GobblinTrackingEvent event;
+
+    while (null != (event = queue.poll())) {
+      GenericRecord record = AvroUtils.overrideNameAndNamespace(event, this.topic, this.namespaceOverride);
+      events.add(Pair.of(buildKey(event), record));
+    }
+
+    if (!events.isEmpty()) {
+      this.pusher.pushKeyValueMessages(events);
+    }
+  }
+
+  private String buildKey(GobblinTrackingEvent event) {
+
+    String key = randomKey;
+    if (this.keys.isPresent()) {
+      StringBuilder keyBuilder = new StringBuilder();
+      for (String keyPart : keys.get()) {
+        Optional value = AvroUtils.getFieldValue(event, keyPart);
+        if (value.isPresent()) {
+          keyBuilder.append(value.get().toString());
+        } else {
+          log.error("{} not found in the GobblinTrackingEvent. Setting key to {}", keyPart, key);
+          keyBuilder = null;
+          break;
+        }
+      }
+
+      key = (keyBuilder == null) ? key : keyBuilder.toString();
+    }
+
+    return key;
+  }
+
+  public static class Factory {
+    /**
+     * Returns a new {@link KeyValueEventObjectReporter.Builder} for {@link KeyValueEventObjectReporter}.
+     * Will automatically add all Context tags to the reporter.
+     *
+     * @param context the {@link MetricContext} to report
+     * @return KafkaReporter builder
+     */
+    public static BuilderImpl forContext(MetricContext context) {
+      return new BuilderImpl(context);
+    }
+  }
+
+  public static class BuilderImpl extends Builder<BuilderImpl> {
+    private BuilderImpl(MetricContext context) {
+      super(context);
+    }
+
+    @Override
+    protected BuilderImpl self() {
+      return this;
+    }
+  }
+
+  public static abstract class Builder<T extends Builder<T>> extends EventReporter.Builder<T> {
```

Comment:
There is no need for further abstraction and generalization. We can simply define `static class Builder extends EventReporter.Builder<Builder>`